README.md
# PLUTO
Primary Layer Updating Tool

## Setup and Building

First, in the PLUTO directory set up the environment by running:

```shell

./setupDevEnvironment.sh
```

To build the docker container:

```shell
npm run build
```

'npm run build' executes the build script, and puts the build into the "./Release" folder.
'npm run docker_build' builds the docker container from this release. Run this before the next
step if you simply want to run the service without developing it. (If you want to develop
jump down to "Starting dev database and running locally" to start a development environment
rather than this release environment.)



## Starting the services

```shell
npm run start_docker
```

It will create a docker volume for the Postgresql database (pgdata), start the database, and then start the web server.

The Postgres database container is listening on port 5432, with user pluto.
The web service container will listen on port 3000, and expects a configuration volume mounted to `/opt/PLUTO/config`.  

The test_config folder at the source root contains a default valid configuration that is mounted to `/opt/PLUTO/config`.
The script will also import the rulesets in `test_config/rulesets` into the database.

It will also start a simple NGINX container to simulate users and authentication, configured from `test_nginx.conf`. 
You can use this by hitting these ports:
  * 8001: no user or group, same as localhost:3000
  * 8002: user="GroupA_user", group="GroupA"
  * 8003: user="GroupB_user", group="GroupB"
  * 8004: user="GroupAdmin_user", group="GroupAdmin", admin=true

It will also start a simulated S3 server to act as a file storage system for PLUTO to download and upload files. See `Using the simulated S3` below.

It will also start a RabbitMQ server which is the job queue that PLUTO workers can be configured to pull jobs from.

It will start a PLUTO web server listening on port 3000.

It will start a PLUTO validation worker that connects to RabbitMQ.

See `test-docker-compose.yml` for the full configuration.

## Stop the services
```shell
npm run stop_docker
```

## Calling the Service

To process a file, POST to `http://localhost:3000/processfile` with a json package. At a minimum, it must specify a ruleset to execute:

```json
{
	"ruleset": "worldCities"
}
```
And can be call via:

```
curl -d '{"ruleset": "sampleRuleset"}' -H "Content-Type: application/json" -X POST http://localhost:3000/processfile
```

If a custom importer is used as part of the [ruleset][ruleset], import configuration (as defined by the importer) can be supplied as part of the `POST`:

```json
{
	"ruleset": "sampleRuleset",
	"import": {
		"file":"http://someurl.com/somefile.csv",
		"auth":"some authentication"
	}
}
```

## Using the simulated S3

Part of the test suite is a simulated S3 using Scality Zenko Cloudserver (scality/s3server on Docker hub), and the 
S3Importer and S3Exporter can be configured to use it. The default access keys and connection info are:

accessKey: accessKey1
secretAccessKey: verySecretKey1
host: localhost or http://localhost <-- ! no ssl/tls
port: 8000

To upload the default data that's used by the sample validations, from a shell in the PLUTO folder:
```bash
# cd data
# ./uploadDevData.sh
```

This script will create the `test` and `testoutput` buckets, and upload the sample data to the s3server container.

The script uses the `s3.js` helper node script to perform these actions.  It can create buckets, upload files to and download files from the s3server container. Run:
```bash
# node s3.js -help
```

To get a list of the available actions.

#### Uploading files via cloudExplorer
Download the latest release from https://github.com/rusher81572/cloudExplorer, unzip and execute CloudExplorer.jar with Java.

Create an account with:
Access Key: accessKey1
Secret Key: verySecretKey1
Host URL: http://localhost
Port: 8000
Name: localhost s3

Save this, then create 2 buckets (from Bucket->Create Bucket): `test` and `testoutput`
Finally, put the data files in `src/examples/data` into the `test` bucket

#### Uploading files via Cyberduck 

Unfortunately, the latest versions of Cyberduck removed the ability to connect to non-https S3 like storage systems, so you'll
need to grab an old version:

Windows: https://update.cyberduck.io/windows/Cyberduck-Installer-4.8.4.19355.exe 
Mac: https://update.cyberduck.io/Cyberduck-4.8.4.19355.zip

Once installed double click on `S3 (HTTP).cyberduckprofile` in the root PLUTO folder to setup a connection, and use the keys
above to access it. The Simulated S3 source uses the bucket `test`, and the Simulated S3 target uses `testoutput`.

## Starting dev database and running the server locally
To run the validator and server locally for dev and debugging purposes, a separate dev environment can be set up. 

The very first time, the containers will need to be built:
```shell
npm run build
docker-compose -f dev-docker-compose.yml build 
```

To start the dev support containers run:

```shell
npm run start_dev_services
```
This will start the dev database on port 6543 and import/update the rulesets from the src/runtime/rulesets folder, a S3 simulator (Scality S3), nginx to emulate different authentication examples, and a RabbitMQ server. See `dev-docker-compose.yml` for more details. 

There are debug configurations also set up in the src folder. To start the web service locally:

### Start the web server
```shell
npm start
```

Or

```shell
npm run start_debug #runs with inspect
```

Or

```shell
cd src
node server/server.js -s server/debugServerConfig.json -v runtime/configs/validatorConfig.json
```

### Start the validator worker
```shell
npm run start_worker
```

Or

```shell
npm run start_worker_debug #runs with inspect
```

Or

```shell
cd src
node validator/queueWorker.js -v runtime/configs/validatorConfig.json
```

### Run the validator manually

```shell
cd src
node validator/validator.js -r CheckDataRulesetConfig -c $PWD/runtime/configs/validatorConfig.json -v $PWD/runtime/rulesets/override.json"
```

or

```shell
cd src
node validator/validator.js -r CheckDataRulesetConfig -c $PWD/runtime/configs/validatorConfig.json -i examples/data/simplemaps-worldcities-basic.csv -o ../results/simplemaps-worldcities-basic.csv.out
```

## Changing the database schema
We're using node-pg-migrate (https://github.com/salsita/node-pg-migrate) to manage migration.  Migration scripts are 
executed as part of the dev database startup and built as part of the pluto_dbloader container.

Docs on the tools are here: https://github.com/salsita/node-pg-migrate

First, a config file needs to be generated for the tool. To generate one against the dev db:

```shell
cd database/dbloader
node configureDatabase.js -v ../../src/runtime/configs/validatorConfig.json
```

which should create dbconfig.json in that folder. To test out the migration:

```shell
../../node_modules/.bin/pg-migrate up -f dbconfig.json
```

To create a new migration script:

```
../../node_modules/.bin/pg-migrate create your-new-migration
```

which will place the script in the database/dbloader/migrations folder.  Follow the docs on how to create new migrations.

## Building the client for development

first time:
```shell
npm install -g ember-cli
```

to build:
```shell
cd src/client
ember build
```

to monitor and build automatically:
```shell
cd src/client
ember server
```

## Python environment
Current Python helper classes and tests are all written in Python 2.7, which needs to be installed. Two packages are required by the PLUTO Python APIs and need to be installed:
 * numpy==1.14.0
 * pandas==0.22.0
 * request==2.18.4

Also add {PLUTO}/src/api to the PYTHONPATH 

## Deploying

After the project is built, in the `Release` folder a `deploy` folder is created that contains some basic info in `readme`, 
a sample config folder, a script to start pluto (create and start the db and start the server), and a sample `Dockerfile` 
as an example on how to extend the pluto container to support plugin dependencies.

Details on configuring the deployed system can be found at [Deployed Configuration][deployedReadme]

## More..

Additional information can be found in the following documents.

[Deployed Configuration][deployedReadme]

[Server][server] <span style="color: red">Needs updating</span>

[Validator][validator]

[Ruleset][ruleset]

[Rules][rules]

[deployedReadme]: deployedReadme.md
[server]: docs/server.md
[validator]: docs/validator.md
[ruleset]: docs/ruleset.md
[rules]: docs/rules.md


CHANGELOG.md
# PLUTO Change Log

## 1.0.17
### New Features
### Bug Fixes and Minor Improvements

## 1.0.16
### New Features
### Bug Fixes and Minor Improvements
* Fixes AUTH-GROUP not set on export. Allows configurable paths for export. Adds mult auth group support to rules (#586)

## 1.0.15
### New Features
* AUTH-GROUP can take a list of groups in a ';' delimited list.  The UI will provide users the ability to select which group to assign as owner. (#581)
### Bug Fixes and Minor Improvements

## 1.0.14
### Bug Fixes and Minor Improvements
* Fixes issue resulting in not all error logs appearing in the UI (#565)
* Display errors in UI for post process tasks (#562)

## 1.0.13
### Bug Fixes and Minor Improvements
* The CheckRowCount rule is now able to check for exact row counts. (#559)
* Choice components that are dynamically populated now support matches when the user types. (#564)
* Post tasks are no longer run when doing a test run. (#563)
* Nginx errors on test are now properly displayed and the Processing icon is removed. (#552)
* Download/upload configuration properties now appear, when permitted. (#556)
* #554 (Parser state is not passed to custom rule) turned out to be a configuration error.
* #562 (Propagate errors to the UI). Errors are now propagated but the component is not yet disabled.

## 1.0.12
### Bug Fixes and Minor Improvements
* Implemented a possible solution for the absent parserState object in Python rules. (#554)
* Stop the "Processing" icon when an error occurs. (Partial fix for #552.)
* Download/Upload properties now appear on Admin page if the user has appropriate authorization. (#556)
* Combo-boxes should be disabled on disabled pages. (#535)

## 1.0.11
### New Features

#### Hide Properties
Added ability to hide properties. (See the bottom of validatorConfig.md.)

#### Show File Details in UI (#442)
The Upload and Download sections of the UI now show the name of the file and where it is.


### Bug Fixes and Minor Improvements
* Fixed "Validation setup not reading default name" (#549)
* "Validation setup not reading default name" (#550) is a duplicate of #549 so was also fixed.
* I can't reproduce "Validation configuration changes do not save after an upload" (#551) and believe
it was an artifact of #549 and therefore fixed.
* I can't reproduce "Processing icon appears on newly added file" (#553) and believe
it was an artifact of #549 and therefore fixed.


## 1.0.10
### New Features

#### Post Upload Tasks
Added a new class of tasks that execute after the file has been successfully uploaded. Uses the same plugin infrastructure as rules, and appear as a list of items below the rules in the edit validation UI.

#### RabbitMQ Queuing of Jobs
Added a job queue (using RabbitMQ) and a validation worker service/container that can be configured to handle the job queue instead of running jobs on the web server.


### Bug Fixes and Minor Improvements
 * Fixed export validation problem related to inconsistent versioning between systems
 * Added hungRunPollingInterval property (can be set in validatorConfig.json) which controls the frequency (6 hours by default) that the database is queried for hung jobs (jobs which are still marked as running after the runMaximumDuration has passed). Found hung jobs are immediately terminated. (#479)
 * Added additional console reports on edit operations. (#507)
 * Jobs are allowed to complete (or timeout) when the server receives a SIGTERM signal. (#512)
 * Added ability to populate UI dropdowns from a JavaScript function on the server
 * Added option to queue validation jobs in RabbitMQ to be run by a separate worker instead of directly on the web server
 * A note was added above the Edit Validation "Save" button when any of the input fields fails validation. (#519)
 * Modified how responses from external processes are received. Should avoid incorrect parsing of JSON responses. (#529)
 * Warn users about unsaved changes when navigating with the browser buttons.
 * Added property value overrides for UI items from the validator config.

## 1.0.9
### New Features

#### Shapefile parsing

PLUTO can now natively parse shapefiles, and all core cvs rules can operate against the shapefile attributes table. Shapefiles can be supplied as a zip, which PLUTO will unzip, validate, and re-zip before uploading, with an option to make the zip contets filenames match the zip filename. It has options to remove all non-shapefile files, and to rename the contents to match the output file name. 

### Bug Fixes and Minor Improvements
 * Added ability for administrators to assign validations to different groups
 * Added a proper label to parsers
 * Added ability to assign custom fields to groups
 * Changed required rules to be per-parser (specifically so validations with 
 * Added config option to restrict access to the admin page to only administrators 
 * Added config option to show a link to the admin page if the user is an administrator
 

## 1.0.8
### New Features

#### Required Rules
Globally required rules can now be specified in the validatorConfig.json, including specific properties. Validations that are missing required rules (or the required config) will fail and new validations will have the required rules automatically added.

### Bug Fixes and Minor Improvements
 * Fixed problems reporting the correct row number when there are an inconsistent number of columns
 * Added the api folder to the PYTHONPATH in the containers, so the PLUTOAPI path can be deprecated
 * Changed the python api to be expected in the PYTHONPATH, as well as adding the api folder to PYTHONPATH in the Docker container
 * Changed rule setup workflow in the validator to support non-parser rules (e.g. unzip a csv)
 * Fixed clone pop-up show the validation id, now shows name
 * Set the Python module Request to version 2.18.4 in the containers
 * Added client validations to the validation general config and base rule properties

startPluto.sh
#!/bin/sh


# cd test_config

# POSTGRES_PORT=$(node -pe "require('./validatorConfig.json').dbPort")
# POSTGRES_USER=$(node -pe "require('./validatorConfig.json').dbUser")
# POSTGRES_PASSWORD=$(node -pe "require('./validatorConfig.json').dbPassword")
# POSTGRES_DATABASE=$(node -pe "require('./validatorConfig.json').dbDatabase")

# cd ..

# docker network create -d bridge --subnet 192.168.0.0/24 --gateway 192.168.0.1 plutonet
# docker volume create pgdata

# docker stop pluto_db
# docker rm pluto_db

# docker run -v pgdata:/var/lib/postgresql/data -p $POSTGRES_PORT:5432 -e POSTGRES_PASSWORD=$POSTGRES_PASSWORD -e POSTGRES_USER=$POSTGRES_USER -e POSTGRES_DB=$POSTGRES_DATABASE --net=plutonet --name pluto_db -d postgres:9.6-alpine
# docker run -v $PWD/test_config:/opt/PLUTO/config --net=plutonet --rm -a stdout -a stderr pluto_dbloader:develop

# docker stop pluto_ws
# docker rm pluto_ws

# docker run -v $PWD/test_config:/opt/PLUTO/config -p 3000:3000 --net=plutonet --name pluto_ws -d pluto:develop

# docker stop pluto_nginx
# docker rm pluto_nginx

# docker run -v $PWD/test_nginx.conf:/etc/nginx/nginx.conf:ro -p 8001:8001 -p 8002:8002 -p 8003:8003 -p 8004:8004 --net=plutonet --name pluto_nginx -d nginx:stable-alpine

# docker run -p 8000:8000 --net=plutonet --name pluto_s3server -d scality/s3server

docker-compose -f test-docker-compose.yml -p pluto_test up -d 